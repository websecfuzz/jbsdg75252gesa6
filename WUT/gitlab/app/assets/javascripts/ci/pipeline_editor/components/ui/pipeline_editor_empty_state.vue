<script>
import { GlButton, GlSprintf, GlEmptyState } from '@gitlab/ui';
import { __ } from '~/locale';
import PipelineEditorFileNav from '~/ci/pipeline_editor/components/file_nav/pipeline_editor_file_nav.vue';

export default {
  components: {
    GlButton,
    GlSprintf,
    GlEmptyState,
    PipelineEditorFileNav,
  },
  i18n: {
    title: __('Optimize your workflow with CI/CD Pipelines'),
    body: __(
      'Create a new %{codeStart}.gitlab-ci.yml%{codeEnd} file at the root of the repository to get started.',
    ),
    btnText: __('Configure pipeline'),
    externalCiNote: __("This project's pipeline configuration is located outside this repository"),
    externalCiInstructions: __(
      'To edit the pipeline configuration, you must go to the project or external site that hosts the file.',
    ),
  },
  inject: {
    emptyStateIllustrationPath: {
      default: '',
    },
    usesExternalConfig: {
      default: false,
      type: Boolean,
      required: false,
    },
  },
  methods: {
    createEmptyConfigFile() {
      this.$emit('createEmptyConfigFile');
    },
  },
};
</script>
<template>
  <div>
    <pipeline-editor-file-nav v-on="$listeners" />
    <gl-empty-state
      v-if="usesExternalConfig"
      :title="$options.i18n.externalCiNote"
      :description="$options.i18n.externalCiInstructions"
      :svg-path="emptyStateIllustrationPath"
    />

    <gl-empty-state v-else :title="$options.i18n.title" :svg-path="emptyStateIllustrationPath">
      <template #description>
        <gl-sprintf :message="$options.i18n.body">
          <template #code="{ content }">
            <code>{{ content }}</code>
          </template>
        </gl-sprintf>
      </template>
      <template #actions>
        <gl-button
          variant="confirm"
          class="gl-mt-3"
          data-testid="create-new-ci-button"
          @click="createEmptyConfigFile"
        >
          {{ $options.i18n.btnText }}
        </gl-button>
      </template>
    </gl-empty-state>
  </div>
</template>
