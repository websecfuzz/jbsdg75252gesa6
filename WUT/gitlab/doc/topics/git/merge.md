---
stage: Create
group: Source Code
info: To determine the technical writer assigned to the Stage/Group associated with this page, see https://handbook.gitlab.com/handbook/product/ux/technical-writing/#assignments
title: Merge your branch into the main branch
---

After you have [created a branch](branch.md), made the required changes, and [committed them locally](commit.md),
you [push your branch](commit.md#send-changes-to-gitlab) and its commits to GitLab.

In the response to the `git push`, GitLab provides a direct link to create the merge request. For example:

```plaintext
...
remote: To create a merge request for my-new-branch, visit:
remote:   https://gitlab.example.com/my-group/my-project/merge_requests/new?merge_request%5Bsource_branch%5D=my-new-branch
```

To get your branch merged into the main branch:

1. Go to the page provided in the link that was provided by Git and
   [create your merge request](../../user/project/merge_requests/creating_merge_requests.md). The merge request's
   **Source branch** is your branch and the **Target branch** should be the main branch.
1. If necessary, have your [merge request reviewed](../../user/project/merge_requests/reviews/_index.md#request-a-review).
1. Have someone [merge your merge request](../../user/project/merge_requests/_index.md#merge-a-merge-request), or merge
   the merge request yourself, depending on your process.

## Related topics

- [Merge requests](../../user/project/merge_requests/_index.md)
- [Merge methods](../../user/project/merge_requests/methods/_index.md)
- [Merge conflicts](../../user/project/merge_requests/conflicts.md)
