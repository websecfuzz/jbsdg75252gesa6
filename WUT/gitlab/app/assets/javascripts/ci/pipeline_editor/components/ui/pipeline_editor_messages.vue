<script>
import { GlAlert } from '@gitlab/ui';
import { getParameterValues, removeParams } from '~/lib/utils/url_utility';
import { __, s__ } from '~/locale';
import {
  COMMIT_FAILURE,
  DEFAULT_FAILURE,
  LOAD_FAILURE_UNKNOWN,
  PIPELINE_FAILURE,
} from '../../constants';
import CodeSnippetAlert from '../code_snippet_alert/code_snippet_alert.vue';
import {
  CODE_SNIPPET_SOURCE_URL_PARAM,
  CODE_SNIPPET_SOURCES,
} from '../code_snippet_alert/constants';

export default {
  components: {
    GlAlert,
    CodeSnippetAlert,
  },

  errors: {
    [COMMIT_FAILURE]: s__('Pipelines|The GitLab CI configuration could not be updated.'),
    [DEFAULT_FAILURE]: __('Something went wrong on our end.'),
    [LOAD_FAILURE_UNKNOWN]: s__('Pipelines|The CI configuration was not loaded, please try again.'),
    [PIPELINE_FAILURE]: s__('Pipelines|There was a problem with loading the pipeline data.'),
  },
  props: {
    failureType: {
      type: String,
      required: false,
      default: null,
    },
    failureReasons: {
      type: Array,
      required: false,
      default: () => [],
    },
    showFailure: {
      type: Boolean,
      required: false,
      default: false,
    },
  },
  data() {
    return {
      codeSnippetCopiedFrom: '',
    };
  },
  computed: {
    failure() {
      const { errors } = this.$options;

      return {
        text: errors[this.failureType] ?? errors[DEFAULT_FAILURE],
        variant: 'danger',
      };
    },
  },
  created() {
    this.parseCodeSnippetSourceParam();
  },
  methods: {
    dismissCodeSnippetAlert() {
      this.codeSnippetCopiedFrom = '';
    },
    dismissFailure() {
      this.$emit('hide-failure');
    },
    parseCodeSnippetSourceParam() {
      const [codeSnippetCopiedFrom] = getParameterValues(CODE_SNIPPET_SOURCE_URL_PARAM);
      if (codeSnippetCopiedFrom && CODE_SNIPPET_SOURCES.includes(codeSnippetCopiedFrom)) {
        this.codeSnippetCopiedFrom = codeSnippetCopiedFrom;
        window.history.replaceState(
          {},
          document.title,
          removeParams([CODE_SNIPPET_SOURCE_URL_PARAM]),
        );
      }
    },
  },
};
</script>

<template>
  <div>
    <code-snippet-alert
      v-if="codeSnippetCopiedFrom"
      :source="codeSnippetCopiedFrom"
      class="gl-mb-5"
      @dismiss="dismissCodeSnippetAlert"
    />
    <gl-alert
      v-if="showFailure"
      :variant="failure.variant"
      class="gl-mb-5"
      @dismiss="dismissFailure"
    >
      {{ failure.text }}
      <ul v-if="failureReasons.length" class="gl-mb-0">
        <li v-for="reason in failureReasons" :key="reason">{{ reason }}</li>
      </ul>
    </gl-alert>
  </div>
</template>
