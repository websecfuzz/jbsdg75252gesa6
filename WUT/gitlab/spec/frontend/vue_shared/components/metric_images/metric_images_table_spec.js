import { GlLink, GlModal } from '@gitlab/ui';
import { shallowMount, mount } from '@vue/test-utils';
import Vue from 'vue';
import merge from 'lodash/merge';
// eslint-disable-next-line no-restricted-imports
import Vuex from 'vuex';
import createStore from '~/vue_shared/components/metric_images/store';
import MetricsImageTable from '~/vue_shared/components/metric_images/metric_images_table.vue';
import MetricImageDetailsModal from '~/vue_shared/components/metric_images/metric_image_details_modal.vue';
import waitForPromises from 'helpers/wait_for_promises';

const defaultProps = {
  id: 1,
  filePath: 'test_file_path',
  filename: 'test_file_name',
};

const mockEvent = { preventDefault: jest.fn() };

Vue.use(Vuex);

describe('Metrics upload item', () => {
  let wrapper;
  let store;

  const mountComponent = (options = {}, mountMethod = mount) => {
    store = createStore();

    wrapper = mountMethod(
      MetricsImageTable,
      merge(
        {
          store,
          propsData: {
            ...defaultProps,
          },
          provide: { canUpdate: true },
        },
        options,
      ),
    );
  };

  const findImageLink = () => wrapper.findComponent(GlLink);
  const findLabelTextSpan = () => wrapper.find('[data-testid="metric-image-label-span"]');
  const findCollapseButton = () => wrapper.find('[data-testid="collapse-button"]');
  const findMetricImageBody = () => wrapper.find('[data-testid="metric-image-body"]');
  const findModal = () => wrapper.findComponent(GlModal);
  const findDeleteButton = () => wrapper.find('[data-testid="delete-button"]');
  const findEditButton = () => wrapper.find('[data-testid="edit-button"]');
  const findImageDetailsModal = () => wrapper.findComponent(MetricImageDetailsModal);
  const closeModal = () => findModal().vm.$emit('hidden');
  const submitModal = () => findModal().vm.$emit('primary', mockEvent);
  const deleteImage = () => findDeleteButton().vm.$emit('click');
  const closeEditModal = () => findImageDetailsModal().vm.$emit('hidden');
  const editImage = () => findEditButton().vm.$emit('click');

  it('render the metrics image component', () => {
    mountComponent({}, shallowMount);

    expect(wrapper.element).toMatchSnapshot();
  });

  it('shows a link with the correct url', () => {
    const testUrl = 'test_url';
    mountComponent({ propsData: { url: testUrl } });

    expect(findImageLink().attributes('href')).toBe(testUrl);
    expect(findImageLink().text()).toBe(defaultProps.filename);
  });

  it('shows a link with the url text, if url text is present', () => {
    const testUrl = 'test_url';
    const testUrlText = 'test_url_text';
    mountComponent({ propsData: { url: testUrl, urlText: testUrlText } });

    expect(findImageLink().attributes('href')).toBe(testUrl);
    expect(findImageLink().text()).toBe(testUrlText);
  });

  it('shows the url text with no url, if no url is present', () => {
    const testUrlText = 'test_url_text';
    mountComponent({ propsData: { urlText: testUrlText } });

    expect(findLabelTextSpan().text()).toBe(testUrlText);
  });

  describe('expand and collapse', () => {
    beforeEach(() => {
      mountComponent();
    });

    it('the card is expanded by default', () => {
      expect(findMetricImageBody().isVisible()).toBe(true);
    });

    it('the card is collapsed when clicked', async () => {
      findCollapseButton().trigger('click');

      await waitForPromises();

      expect(findMetricImageBody().isVisible()).toBe(false);
    });
  });

  describe('delete functionality', () => {
    it('should open the delete modal when clicked', async () => {
      mountComponent({ stubs: { GlModal: true } });

      deleteImage();

      await waitForPromises();

      expect(findModal().attributes('visible')).toBe('true');
    });

    describe('when the modal is open', () => {
      beforeEach(() => {
        mountComponent(
          {
            data() {
              return { modalVisible: true };
            },
          },
          shallowMount,
        );
      });

      it('should close the modal when cancelled', async () => {
        closeModal();

        await waitForPromises();
        expect(findModal().attributes('visible')).toBeUndefined();
      });

      it('should delete the image when selected', async () => {
        const dispatchSpy = jest.spyOn(store, 'dispatch').mockImplementation(jest.fn());

        submitModal();

        await waitForPromises();

        expect(dispatchSpy).toHaveBeenCalledWith('deleteImage', defaultProps.id);
      });
    });

    describe('canUpdate permission', () => {
      it('delete button is hidden when user lacks update permissions', () => {
        mountComponent({ provide: { canUpdate: false } });

        expect(findDeleteButton().exists()).toBe(false);
      });
    });
  });

  describe('edit functionality', () => {
    it('should open the metric image details dialog when clicked', async () => {
      mountComponent({ stubs: { MetricImageDetailsModal: true } });

      editImage();

      await waitForPromises();

      expect(findImageDetailsModal().attributes('visible')).toBe('true');
    });

    it('should close the metric image details dialog when cancelled', async () => {
      mountComponent({
        data() {
          return { editModalVisible: true };
        },
        stubs: { MetricImageDetailsModal: true },
      });

      closeEditModal();

      await waitForPromises();

      expect(findImageDetailsModal().attributes('visible')).toBeUndefined();
    });
  });
});
