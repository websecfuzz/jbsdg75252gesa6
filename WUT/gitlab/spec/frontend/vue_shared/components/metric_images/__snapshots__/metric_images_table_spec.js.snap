// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metrics upload item render the metrics image component 1`] = `
<gl-card-stub
  bodyclass="[object Object]"
  class="gl-mb-5"
  footerclass=""
  headerclass="gl-flex,gl-items-center,[object Object]"
>
  <gl-modal-stub
    actioncancel="[object Object]"
    actionprimary="[object Object]"
    arialabel=""
    body-class="!gl-pb-0 !gl-min-h-6"
    dismisslabel="Close"
    modalclass=""
    modalid="delete-metric-modal"
    size="sm"
    titletag="h4"
  >
    <p>
      Are you sure you wish to delete this image?
    </p>
  </gl-modal-stub>
  <metric-image-details-modal-stub
    edit="true"
    filename="test_file_name"
    imagefiles=""
    imageid="1"
  />
  <div
    class="gl-flex gl-flex-col"
    data-testid="metric-image-body"
  >
    <img
      class="gl-max-w-full gl-self-center"
      src="test_file_path"
    />
  </div>
</gl-card-stub>
`;
